### ISSUE TITLE
V3 powerwall SOC and energy remaining drop to 0

### ISSUE THREAD

**Issue by purcell-lab** [alandtse/tesla#296]
## Version of the custom_component

3.0.0

## Describe the bug

Raising an issue here to highlight an upstream bug that powerwall remaining energy and battery % regularly drop to 0 and then back to their actual values.

![Screenshot_20221023-122435](https://user-images.githubusercontent.com/79175134/197370192-791f692c-8cef-4eef-b378-030d0a59d8c6.png)

The Tesla API is the one who is reporting this drop to 0 so it is an upstream issue.

Workarounds include using an available: statement in template sensors or some type of averaging template.

**Comment by purcell-lab** [comment]
This workaround doesn't handle the legitimate 0 case, but is better than the stock item.

```yaml
 - platform: template
  sensors:
    filtered_powerwall_soc:
      value_template: "{{states('sensor.home_energy_gateway_battery')}}"
      availability_template: "{{states('sensor.home_energy_gateway_battery')|int(0) != 0}}"
      unit_of_measurement: "%"

    filtered_powerwall_battery_remaining: 
      value_template: "{{ states('sensor.home_energy_gateway_battery_remaining') }}"
      availability_template: "{{states('sensor.home_energy_gateway_battery_remaining')|int(0) != 0}}"
      unit_of_measurement: Wh
```

**Comment by stevesreed** [comment]
I'm having this issue as well. It is still present in 3.0.2

**Comment by Ottmar0815** [comment]
Here also. It's still present.

**Comment by purcell-lab** [comment]
> I have my own script that polls the local (gateway) API every 20 seconds and retrieves data from there, including the SoC (battery level %).

I agree this issue doesn't appear on the local API.

This integration uses the cloud API, feel free to poll the cloud API and let us know if you receive zeros. I have and I do get zeros as per the issue report.

Zero is a valid response but only when the battery is fully discharged. Interestingly the local API reports 5% when fully discharged.

**Comment by jaymunro** [comment]
See my comment in #475 that points out that the power reading timestamp does not match the request time but varies quite randomly. Does this suggest that the API call is missing a parameter to specify the request time?
https://github.com/alandtse/tesla/issues/475#issuecomment-1376669201

**Comment by ChrisJ60** [comment]
Yes, 0% is valid if the battery is fully discharged but a cursory glance at the returned values over time shows this is not the case here though. So these sudden drops to zero are bogus.

Using the Cloud API is probably unwise, at least to read data (maybe it is necessary for control actions, I don't know). The response time for the local API is much better (of course) and also it does not depend on Internet connectivity. Win-win.

The Powerwall reserves 5% pf capacity at the low end as a buffer to avoid true 100% discharge of the batteries and the local API reports SoC as a %age of the total capacity rather than the usable capacity. So SoC calculations based on the local API need to take that into account (my utility does this, it isn't hard).

Chris

On 25 May 2023, at 21:05, purcell-lab ***@***.***> wrote:


I have my own script that polls the local (gateway) API every 20 seconds and retrieves data from there, including the SoC (battery level %).

I agree this issue doesn't appear on the local API.

This integration uses the cloud API, feel free to poll the cloud API and let us know if you receive zeros. I have and I do get zeros as per the issue report.

Zero is a valid response but only when the battery is fully discharged. Interestingly the local API reports 5% when fully discharged.

**Comment by jaymunro** [comment]
This looks to be resolved by #724

**Comment by purcell-lab** [comment]
Very pleased to see this fixed.

Thanks all.
